src/lib/plugin.js

```diff
@@ -274,22 +278,31 @@ module.exports = class PluginVirtual extends EventEmitter2 {
     const now = new Date()
 
     const that = this
-    setTimeout(co.wrap(function * () {
-      debug('going to time out ' + transferId)
+    setTimeout(
+      co.wrap(this._expireTransfer).bind(this, transferId),
+      (expiry - now))
+  }
 
-      const packaged = yield that._transfers._getPackaged(transferId)
+  * _expireTransfer (transferId) {
+    debug('checking time out on ' + transferId)
 
-      // don't cancel again if it's already cancelled
-      if (!(yield that._transfers.cancel(transferId))) {
-        debug(transferId + ' has already cancelled')
+    const packaged = yield this._transfers._getPackaged(transferId)
+
+    // don't cancel again if it's already cancelled/executed
+    try {
+      if (!(yield this._transfers.cancel(transferId))) {
+        debug(transferId + ' is already cancelled')
```

Comment:
in my transfer state change function, it returns false if the state change has already been done (to make idempotence easier) and throws an error if the state is different (eg. the state is executed and you're trying to cancel)